rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+
+
+## Design overview
+
+### End-to-end workflow overview
+
+Case where a user has a single preprocessing layer to do image normalization.
+
+```python
+normalization = keras.layers.Normalization(axis=-1)
+normalization.adapt(data_sample)
+
+model = keras.Sequential([
+    normalization,
+    keras.applications.ResNet50(weights=None),
+])
+model.fit(data, targets, epochs=10)
+```
+
+Case where a user has a single preprocessing layer to do text vectorization where each input sample is encoded as a sequence of word indices.
+
+```python
+vectorization = keras.layers.TextVectorization(mode='int')
+vectorization.adapt(data_sample)
+
+model = keras.Sequential([
+    vectorization,
+    keras.layers.Embedding(128),  # The number of int indices is not specified since it is inferred.
+    keras.layers.LSTM(32),
+    keras.layers.Dense(10, activation='softmax'),
+])
+model.fit(data, targets, epochs=10)
+```
+
+Case where a user has a single preprocessing layer to do text vectorization where each input sample is encoded as a dense vector of TF-IDF scores.
+
+```python
+vectorization = keras.layers.TextVectorization(mode='tfidf')
+vectorization.adapt(data_sample)
+
+model = keras.Sequential([
+    vectorization,
+    keras.layers.Dense(10, activation='softmax'),
+])
+model.fit(data, targets, epochs=10)
+```
+
+Case where a user chains a a normalization step with a discretization step.
+
+```python
+normalization = keras.layers.Normalization()
+discretization = keras.layers.Discretization()
+preprocessing_stage = keras.layers.PreprocessingStage([normalization,
+                                                       discretization])
+preprocessing_stage.adapt(data_sample)
+
+model = keras.Sequential([
+    preprocessing_stage,
+    keras.layers.Dense(10, activation='softmax'),
+])
+model.fit(data, targets, epochs=10)
+```
+
+
+### Base class: `PreprocessingLayer`
+
+All preprocessing layers inherit from a base class: `PreprocessingLayer`, which itself inherits from `Layer`.
+
+This class presents a few key differences compared to regular layers:
+
+**Separate training mechanism**
+
+The internal state of a `PreprocessingLayer` is not affected by backpropagation: all of its weights are non-trainable. A `PreprocessingLayer` has to be trained in a separate step, as follow:
+
+```python
+preprocessing_layer.adapt(data_sample)
+```
+
+**Possible non-differentiability**
+
+Processing layers extend Keras by allowing preprocessing to be part of the model. Unlike existing layers, these computations are not always differentiable, e.g. both `Discretize` and `VectorizeText` are non-differentiable.
+
+As a result, all preprocessing layers are treated as frozen when used as part of a model. In addition, if a non-differentiable layer is used in the middle of a model (rather than at the start), the model will raise an exception related to differentiability when trying to compute gradients (e.g. as part of `fit`).
```

Comment:
I am assuming here that this validation will be at compile time.

Reply:
Probably at the start of `fit`/etc. You don't need gradients for `evaluate` or `predict` so raising in `compile` is too early (we don't actually query gradients in `compile` today).